compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -1555,6 +1555,38 @@ public boolean isAssignableToFiniteType(BType type,
                         if (((BLangLiteral) memberLiteral).value == null) {
                             return literalExpr.value == null;
                         } else {
+                            // Check the literal value and the literal's type's tag (There will be different tags for
+                            // decimal and float types)
+                            return (((BLangLiteral) memberLiteral).value.equals(literalExpr.value) &&
+                                    ((BLangLiteral) memberLiteral).type.tag == literalExpr.typeTag);
+                        }
+                    });
+            return foundMember;
+        }
+        return false;
+    }
+
+    /**
+     * This method checks the typeTag of the literal along with the literal value to check the assignability.
+     * Unlike {@link #isAssignableToFiniteType(BType, BLangLiteral)} method, here we check the literal's typeTag
+     * rather than checking the literal's type's tag. This allows us to assign a float literal directly to a finite
+     * type that consists of a corresponding decimal value. Use this method when we assign a BLangLiteral directly to
+     * a finite type.
+     *
+     * @param type        expected BType
+     * @param literalExpr literal expression
+     * @return returns true if assignable, false otherwise
+     */
+    boolean isLiteralAssignableToFiniteType(BType type, BLangLiteral literalExpr) {
+        if (type.tag == TypeTags.FINITE) {
```

Comment:
The reason why we need two methods is, one method checks the `literal.type.tag` when checking the assignability and the latter uses `literal.typeTag`. Consider the following scenario, where we have a finite type made up of decimal constants and we try to assign a simple floating point literal to a variable of user-defined finite type. 

```ballerina
const decimal CONST1 = 9.23;
const decimal CONST2 = 21.4;

type DEC CONST1|CONST2;

public function main() {
    DEC d1 = 21.4; // Case 1: Expected to be successful

    float f1 = 21.4;
    DEC d2 = f1; // Case 2: Expected to fail at compile time
}
```
As shown above, we need to support assigning a simple floating point literal to the finite type made up of decimal constants. However, simple float literal 21.4 would have its `literal.type.tag`
as float's tag and the finite type `valuespace` values would have `literal.type.tag` of decimal's tag as it is made up of decimal constants. By considering the `literal.typeTag` when checking the assignability we can solve this problem as both decimal and float literals expected to have the same `literal.typeTag`. 

At the same time, we need to fail the case 2 at compile time as it tries to assign a float variable to the finite type made up of decimal types. Hence, in this case, we need to consider the `literal.type.tag` when checking the assignability as it would be different values for decimal and float types.

This is the reason why we need two different methods to handle both the above cases.